junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,525 @@
+ *               annotations = List(new Before)),
+ *           new JUnitMethodMetadata(name = "bar", id = "1",
+ *               annotations = List(new Test)),
+ *           new JUnitMethodMetadata(name = "baz", id = "2",
+ *               annotations = List(new Test, new Ignore("baz not implemented yet")))
+ *       ),
+ *       moduleMethods(
+ *           new JUnitMethodMetadata(name = "beforeClass", id = "0",
+ *               annotations = List(new BeforeClass)))
+ *     )
+ *   }
+ *   def newInstance(): ScalaJSJUnitTest = new Foo()
+ *   def scalajs$junit$invoke(methodId: String): Unit = {
+ *     if (methodId = "0") Foo.beforeClass()
+ *     else throw new NoSuchMethodException(methodId)
+ *   }
+ * }
+ *
+ * The test framework will identify Foo$scalajs$junit$hook as a test module
+ * because it extends JUnitTestMetadata. It will know which methods to run based
+ * on the info returned by Foo$scalajs$junit$hook.metadata,
+ * it will create new test instances using Foo$scalajs$junit$hook.newInstance()
+ * and it will invoke test methods using scalajs$junit$invoke methods on the
+ * module or instance.
+ */
+class ScalaJSJUnitPlugin(val global: Global) extends NscPlugin {
+  selfPlugin =>
+
+  val name: String = "Scala.js JUnit plugin"
+
+  val components: List[NscPluginComponent] =
+    List(ScalaJSJUnitPluginComponent)
+
+  val description: String = "Makes JUnit test classes invokable in Scala.js"
+
+  object ScalaJSJUnitPluginComponent
+    extends plugins.PluginComponent with transform.Transform {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = ScalaJSJUnitPlugin.this.global
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends global.Transformer {
+
+      import global._
+      import rootMirror.getRequiredClass
+
+      private val testAnnotationType =
+        getRequiredClass("org.junit.Test").toType
+
+      private val fixMethodOrderAnnotationType =
+        getRequiredClass("org.junit.FixMethodOrder").toType
+
+      private val annotationWhiteList = List(
+        testAnnotationType,
+        getRequiredClass("org.junit.Before").toType,
+        getRequiredClass("org.junit.After").toType,
+        getRequiredClass("org.junit.BeforeClass").toType,
+        getRequiredClass("org.junit.AfterClass").toType,
+        getRequiredClass("org.junit.Ignore").toType
+      )
+
+      private val scalaJSJUnitTestType =
+        getRequiredClass("org.scalajs.junit.ScalaJSJUnitTest").toType
+
+      private val jUnitClassMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitClassMetadata").toType
+
+      private val jUnitTestMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitTestMetadata").toType
```

Comment:
It is very uncomment to work directly with types when within scalac. You should use symbols directly whenever you are working with class types. They have sane equality and are lower overhead (because they only need to handle class types, not *any* kind of type).